server/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java

```diff
@@ -79,12 +81,20 @@ public DeprecationLoggerBuilder deprecate(final String key, final String msg, fi
 
     public class DeprecationLoggerBuilder {
 
-        public DeprecationLoggerBuilder withDeprecation(String key, String msg, Object[] params) {
+        public DeprecationLoggerBuilder withDeprecation(String key, String msg, Object... params) {
             String opaqueId = HeaderWarning.getXOpaqueId();
             ESLogMessage deprecationMessage = DeprecatedMessage.of(opaqueId, msg, params);
             deprecationLogger.throttleLogAndAddWarning(key, deprecationMessage);
             return this;
         }
 
+        public DeprecationLoggerBuilder compatibleApiWarning(String key, String msg, Object... params) {
+            String opaqueId = HeaderWarning.getXOpaqueId();
+            ESLogMessage deprecationMessage = DeprecatedMessage.compatibleDeprecationMessage(opaqueId, msg, params);
+            compatibleLogger.throttleLogAndAddWarning(key, deprecationMessage);
```

Comment:
can't remember what we agreed. Do we throttle compatible API logs?

Reply:
yes throttled, however, I think the they key needs to be more contextual. i.e. rather then just a "get_with_types" key, it needs to be the actual resolved message emitted, which should include the per request elements such as the path parameters. Which means that it will have ALOT more cache misses and may want to separate the backing cache store from the deprecation logger.

Reply:
agree. We could either create a second cache (same LRU set) or just expand it to make it bigger. Not sure we could make the size of that cache configurable. This would allow users to estimate how many `x-opaque-id` unique users will they have (we would then multiply that number by number of compatible APIs we have).

When it comes to the key itself, and if we want to use `StackWalker` we need to make sure this is properly abstracted so that it won't be accidentally back ported to 7.x where jdk 1.8 is still supported. 
I guess for the time being we could simply just resolve the formatted message as you suggest.

Reply:
for the record. We discussed this as I was concerned that expanding path parameters into logs would make throttling not be as effective. Every log line with a key containing a different type would not throttle.
However there is a plus side to this - a user might easier identify places in his codebase where he uses that type. For instance if his app has that type "hardcoded" somewhere.

We are considering if we could have a debug level log line with exact path like `myIndex/myType/` and a warn level with a path being in a form of `{index}/{type}`

For the time being we will stay with the more verbose expanded version and will revisit if this becomes a problem

Reply:
I am still not sure on logs where an id will be used. For this kind of endpoints there would be far too many logs. i.e. `/{index}/{type}/{id}`